chore(learn): Add details about math functions, improve intro (#28949)
mdn/content: files/en-us/learn/css/building_blocks/values_and_units/index.md

{{LearnSidebar}}{{PreviousMenuNext("Learn/CSS/Building_blocks/Overflowing_content", "Learn/CSS/Building_blocks/Sizing_items_in_CSS", "Learn/CSS/Building_blocks")}}

> The syntax of CSS uses [declarations](/en-US/docs/Web/CSS/Syntax#css_declarations) made of properties and values to create rules.

Reviewer: Maybe I'm just being ultra nitpicky, but this intro rewrite seemed a bit off to me. It felt like it needed tightening up a bit, and the use of the word "something" sounded a bit vague and weird. I also felt like the link to the property reference should be a note below rather than a part of the paragraph. It feel tangential to the story being told.

How about something like the following?

---

CSS rules contain [declarations](/en-US/docs/Web/CSS/Syntax#css_declarations), which in turn are composed of properties and values. Each property used in CSS has a **value type** that describes what kind of values it is allowed to have. In this lesson, we will take a look at some of the most frequently used value types, what they are, and how they work.

> **Note:** Each [CSS property page](/en-US/docs/Web/CSS/Reference#index) has a syntax section that lists its value type.

Author: No that's much better. Added in 0b7e13af74e865dc3448a063edadd5a3fd0b61f5 with a minor tweak to the note.

## Functions

In programming, a function is a piece of code that does a specific task.
Functions are useful because you can write code once, then reuse it many times instead of writing the same logic over and over.

> A lot of programming languages support functions and the convenient thing is that many languages come with built-in functions for common tasks so you don't have to write them yourself from scratch.

Reviewer: ```suggestion
Most programming languages not only support functions but also come with convenient built-in functions for common tasks so you don't have to write them yourself from scratch.
```

> CSS also has [functions](/en-US/docs/Web/CSS/CSS_Functions) (which you might see written as _functional notation_), and they work in a similar way as in other languages.

Reviewer: ```suggestion
CSS also has [functions](/en-US/docs/Web/CSS/CSS_Functions), which work in a similar way to functions in other languages.
```

Slightly awkward phrasing, plus I thought the mention of functional notation hindered more than it helped for the beginner audience - made it sound like it was something else special you need to know, whereas really it is just the way they are written.

In fact, we've already seen CSS functions in the [Color](#color) section above with [`rgb()`](/en-US/docs/Web/CSS/color_value#rgb_function) and [`hsl()`](/en-US/docs/Web/CSS/color_value#hsl_function) functions.

Aside from applying colors, you can use functions in CSS to do a lot of other things.
For example [Transform functions](/en-US/docs/Web/CSS/CSS_Functions#transform_functions) are a common way to move, rotate, and scale elements on a page.
You might see [`translate()`](/en-US/docs/Web/CSS/transform-function/translate) for moving something horizontally or vertically, [`rotate()`](/en-US/docs/Web/CSS/transform-function/rotate) to rotate something, or [`scale()`](/en-US/docs/Web/CSS/transform-function/scale) to make something bigger or smaller.

### Math functions

When you are creating styles for a project, you will probably start off with numbers like `300px` for lengths or `200ms` for durations.
If you want to have these values change based on other values, you will need to do some math.

> You could calculate the percentage of a value, add or subtract a number from another number, then update your CSS with the result.

Reviewer: ```suggestion
You could calculate the percentage of a value or add a number to another number, then update your CSS with the result.
```

> Luckily, CSS has support for [Math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions) which helps us perform calculations instead of writing out the numbers ourselves or even doing the math in JavaScript.

Reviewer: ```suggestion
CSS has support for [Math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions), which allow us to perform calculations instead of relying on static values or doing the math in JavaScript.
```

One of the most common math functions is [`calc()`](/en-US/docs/Web/CSS/calc) which lets you do operations like addition, subtraction, multiplication, and division.

> Let's say you want values in your CSS that the browser should work out for you when someone is viewing your page.

Reviewer: I'd just delete this line. Feels like it labors the point a little too much, plus you are mixing "you" and "we".

> For example, we want to set the width of an element to be 20% of its parent container and add 100px to it.

Reviewer: ```suggestion
For example, let's say we want to set the width of an element to be 20% of its parent container plus 100px.
```

> We might not be able to do this calculation beforehand if we don't know the width of the parent container if the parent uses a percentage width or a relative unit like `em` or `rem`.

Reviewer: ```suggestion
We can't specify this width with a static value — if the parent uses a percentage width (or a relative unit like `em` or `rem`) then it will vary depending on the context it is used in, and other factors such as the user's device or browser window width.
```

> In our example, we can use `calc()` to calculate the width of the element and set it to be 20% of the parent container plus 100px.

Reviewer: ```suggestion
However, we can use `calc()` to set the width of the element to be 20% of its parent container plus 100px.
```

The 20% is based on the width of the parent container (`.wrapper`) and if that width changes, the calculation will change too:

{{EmbedGHLiveSample("css-examples/learn/values-units/calc.html", '100%', 500)}}

> There are many other math functions that you can use in CSS, such as [`min()`](/en-US/docs/Web/CSS/min), [`max()`](/en-US/docs/Web/CSS/max), and [`clamp()`](/en-US/docs/Web/CSS/clamp) which let you pick the smallest, largest, or middle value from a set of values.

Reviewer: ```suggestion
There are many other math functions that you can use in CSS, such as [`min()`](/en-US/docs/Web/CSS/min), [`max()`](/en-US/docs/Web/CSS/max), and [`clamp()`](/en-US/docs/Web/CSS/clamp); respectively these let you pick the smallest, largest, or middle value from a set of values.
```

> You can also use [Trigonometric functions](/en-US/docs/Web/CSS/CSS_Functions#trigonometric_functions) like [`sin()`](/en-US/docs/Web/CSS/sin), [`cos()`](/en-US/docs/Web/CSS/cos), and [`tan()`](/en-US/docs/Web/CSS/tan) to calculate angles for rotating elements around a point, or even choosing colors that take a [hue angle](/en-US/docs/Web/CSS/hue) as a parameter.

Reviewer: ```suggestion
You can also use [Trigonometric functions](/en-US/docs/Web/CSS/CSS_Functions#trigonometric_functions) like [`sin()`](/en-US/docs/Web/CSS/sin), [`cos()`](/en-US/docs/Web/CSS/cos), and [`tan()`](/en-US/docs/Web/CSS/tan) to calculate angles for rotating elements around a point, or choose colors that take a [hue angle](/en-US/docs/Web/CSS/hue) as a parameter.
```

> [Exponential functions](/en-US/docs/Web/CSS/CSS_Functions#exponential_functions) might also be used for animations, transitions, or having a very specific control over how something moves and looks.

Reviewer: ```suggestion
[Exponential functions](/en-US/docs/Web/CSS/CSS_Functions#exponential_functions) might also be used for animations and transitions, when you require very specific control over how something moves and looks.
```

> Knowing about these functions is useful so you recognize them when you see them, and if you feel like using them in your projects, they will help you avoid writing custom or repetitive code that you can do with regular CSS.

Reviewer: ```suggestion
Knowing about CSS functions is useful so you recognize them when you see them. You should start experimenting with them in your projects — they will help you avoid writing custom or repetitive code to achieve results that you can get with regular CSS.
```
